WFLY-9037 Cluster topology updates are lossy

pferraro opened the pull request:
https://issues.jboss.org/browse/WFLY-9037

~~N.B. All ejb client clustering tests are now working, with the exception of the authentication one, which has more to do with my inability to setup the test correctly.
We can merge a fix for that later.~~
Farah has provided a workaround, now incorporated as a separate test using a global authentication context.

pferraro committed:
Add Group.isLocal() so that the EJB subsystem can easily distinguish a local group from a clustered group.

pferraro committed:
Use Group.isLocal() to determine Affinity.

pferraro committed:
WFLY-9037 Provide client mappings registry eagerly to AssociationService, instead of lazily via RegistryCollector.
Register single listener with Registry, and delegate events to each connection-specific topology listener.
Ensure initial topology is set before attempting to process topology changes via Registry listener.

pferraro committed:
Make AssociationService on-demand, and its dependents either passive or on-demand.

pferraro committed:
Split RemoteFailoverTestCase into separate tests. Disable reuseForks, absent another mechanism to isolate ejb client context between tests.

pferraro committed:
Remove obsolete optional dependencies.

pferraro committed:
Remove ControlledProcessStateService logic - since this service is now started only when demanded by a deployment containing an EJB with a remote view.

kabir commented:
@pferraro Unfortunately we seem to have a failure in https://ci.wildfly.org/viewLog.html?buildId=66718&tab=buildResultsDiv&buildTypeId=WFPR

pferraro commented:
@kabir That run was outdated - I've kicked off another.

kabir commented:
@pferraro Still seems problematic, unless the runs still are stale. For linux I've kicked off builds #12863-65 and for windows #7896-98.

pferraro commented:
@kabir The linux runs all passed like a champ.  The windows builds are failing all over the place and in the weirdest of places.  The downstream windows builds look fine.

dmlloyd commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3RemoteServiceAdd.java:
```diff
@@ -123,7 +123,7 @@ void installRuntimeServices(final OperationContext context, final ModelNode mode
                 .addDependency(TxnServices.JBOSS_TXN_REMOTE_TRANSACTION_SERVICE, RemotingTransactionService.class, ejbRemoteConnectorService.getRemotingTransactionServiceInjector())
                 .addDependency(ControlledProcessStateService.SERVICE_NAME, ControlledProcessStateService.class, ejbRemoteConnectorService.getControlledProcessStateServiceInjector())
-                .setInitialMode(ServiceController.Mode.ACTIVE);
+                .setInitialMode(ServiceController.Mode.ON_DEMAND);
```
What is the purpose of doing this?

pferraro replied:
Because there is no reason for it to be started ACTIVEly.  This service is only needed by an EJB deployment with a remote view, and thus should only start when needed by that deployment.
Additionally, this service will now force the Registry to start, and we don't need to start any clustering services unless they are needed by a deployment.

dmlloyd commented on ejb3/src/main/java/org/jboss/as/ejb3/remote/AssociationImpl.java:
```diff
@@ -73,23 +73,59 @@
  * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
  */
-final class AssociationImpl implements Association {
+final class AssociationImpl implements Association, Registry.Listener<String, List<ClientMapping>>, AutoCloseable {
```
Why implement the listener interface directly?

dmlloyd replied:
Are there functional changes hidden in here or is it just a code-shuffle refactor?

pferraro replied:
"Why implement the listener interface directly?"
Because this object listens for registry events. 

"Are there functional changes hidden in here or is it just a code-shuffle refactor?"
In this class specifically?  The functional changes are:
* The Association object operates directly upon a specific Registry instead of one lazily supplied by the RegistryCollector.
* Replace Registry listener per connection with a single Registry listener for the Association object.  Registry events are queued and processed by a single thread per listener to maintain ordering, thus using a listener per connection is not appropriate.  The Association object maintains a separate collection of connection-specific ClusterTopologyListeners, to which Registry events are delegated.
* Ensure initial topology is set before attempting to process topology changes via Registry listener (see comments above the corresponding synchronization blocks).

dmlloyd replied:
I'd prefer to stick with composition for the listener instead of inheritance in this particular case.  In the general case, it's important to separate structure refactors from logic changes in PRs because otherwise it's not clear what the actual change is.

dmlloyd replied:
I don't think that the fact that events are queued sequentially and the fact that they are per-connection are logically connected in any way (each connection independently has a sequential view, and each connection is separately asynchronous in terms of when it receives/processes events), thus from that perspective it doesn't seem to matter whether there is a listener per connection or a single listener which is delegated to many connections.

Also it's worth noting that there is exactly one consumer of the Registry API, so it's reasonable to expect that the API matches the needs of the consumer as closely as possible, as opposed to requiring a second layer of listener tracking.

pferraro replied:
"I'd prefer to stick with composition for the listener instead of inheritance in this particular case."
OK.

"I don't think that the fact that events are queued sequentially and the fact that they are per-connection are logically connected in any way (each connection independently has a sequential view, and each connection is separately asynchronous in terms of when it receives/processes events), thus from that perspective it doesn't seem to matter whether there is a listener per connection or a single listener which is delegated to many connections."
Each Registry listener is allocated a blocking queue and single thread for processing topology change events, as Registry.Listeners are contractually non-interfering.  Given that the number of connections can be quite large, it doesn't make sense to register a Registry.Listener per connection.

"Also it's worth noting that there is exactly one consumer of the Registry API."
On the contrary, each Undertow server uses a Registry to lookup routing identifiers for each cluster member.  Additionally, as part of the public clustering API, the Registry API is exposed to deployments.  The singleton domain controller will likely be another consumer.

dmlloyd commented on ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/StatefulComponentCreateServiceFactory.java:
```diff
@@ -60,13 +57,6 @@ public void configureDependency(ServiceBuilder<?> serviceBuilder, StatefulSessio
         configuration.getCreateDependencies().add(new DependencyConfigurator<StatefulSessionComponentCreateService>() {
             @Override
-            public void configureDependency(ServiceBuilder<?> builder, StatefulSessionComponentCreateService service) {
-                builder.addDependency(DependencyType.OPTIONAL, ClientMappingsRegistryBuilder.SERVICE_NAME);
```
Always a +1 for removing `DependencyType.OPTIONAL`...

dmlloyd commented on ejb3/src/main/java/org/jboss/as/ejb3/remote/EJBRemoteConnectorService.java:
```diff
@@ -75,35 +70,14 @@ public void start(StartContext context) throws StartException {
             associationService.getAssociation(),
             remotingTransactionServiceInjectedValue.getValue()
         );
-        final ControlledProcessStateService processStateService = controlledProcessStateServiceInjectedValue.getValue();
```
Are you sure this is safe?  A service can be `ON_DEMAND` and still start up eagerly.

pferraro replied:
As of this PR, the only EJBRemoteConnectorService dependent is an EJB component with a Remote interface.  See EJBComponentDescription.setupRemoteView(...).  This may have made since when the service was started actively, but I don't see why it would be necessary now.

pferraro committed:
Don't use deprecated jndi initial context factory.

pferraro committed:
Change AssociationService and dependents to start LAZYly to retain ClusterTopologyListeners listeners for lifecycle of remoting channel.
This isn't ideal, and hopefully temporary, but is sufficient to workaround race conditions between the channel and AssociationService lifecycles.

kabir commented:
Retest this please

dmlloyd reviewed (approved):
I'm OK with this code as of dbe280f0499d58f1330ab882e456901b54fff26b .

kabir commented:
Retest this please
